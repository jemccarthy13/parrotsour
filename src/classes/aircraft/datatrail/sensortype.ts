export enum SensorType {
  RAW,
  ARROW,
}
