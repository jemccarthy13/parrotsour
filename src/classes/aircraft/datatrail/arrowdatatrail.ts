import { PaintBrush } from "../../../canvas/draw/paintbrush"
import { headingToRadians, PIXELS_TO_NM } from "../../../utils/psmath"
import { Point } from "../../point"
import { IDMatrix } from "../id"
import { DataTrail } from "./datatrail"

export class ArrowDataTrail extends DataTrail {
  public static LEN_TRAIL = 5
  getCenterOfMass(heading: number): Point {
    const vector = headingToRadians(heading)

    const dist = ArrowDataTrail.LEN_TRAIL * PIXELS_TO_NM

    const startPos = this.getStartPos()
    return new Point(
      Math.floor(startPos.x + 1.2 * dist * Math.cos(vector.radians)),
      Math.floor(startPos.y + 1.2 * dist * -Math.sin(vector.radians))
    )
  }

  draw(heading: number, id: IDMatrix): void {
    const ctx = PaintBrush.getContext()
    ctx.lineWidth = 1
    ctx.fillStyle = id

    let endx = 0
    let endy = 0

    const vector = headingToRadians(heading)

    const dist = ArrowDataTrail.LEN_TRAIL * PIXELS_TO_NM

    const startPos = this.getStartPos()
    const startx = startPos.x
    const starty = startPos.y

    endy = starty + dist * -Math.sin(vector.radians)
    endx = startx + dist * Math.cos(vector.radians)

    ctx.beginPath()
    ctx.moveTo(startx, starty)
    ctx.lineTo(endx, endy)

    const heady: number =
      endy +
      ArrowDataTrail.LEN_TRAIL *
        0.4 *
        PIXELS_TO_NM *
        -Math.sin(vector.headAngle)
    const headx: number =
      endx +
      ArrowDataTrail.LEN_TRAIL * 0.4 * PIXELS_TO_NM * Math.cos(vector.headAngle)

    ctx.lineTo(headx, heady)

    ctx.strokeStyle = id
    ctx.stroke()
    ctx.stroke()
    ctx.stroke()
  }
}
