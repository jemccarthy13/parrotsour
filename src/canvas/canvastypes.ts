import { Point } from "classes/point"
import { FORMAT } from "../classes/supportedformats"
import { SensorType } from "classes/aircraft/datatrail/sensortype"
import { AircraftGroup } from "classes/groups/group"

export enum BlueInThe {
  NORTH,
  SOUTH,
  EAST,
  WEST,
}

export class FightAxis {
  public static isNS(orientation: BlueInThe): boolean {
    return orientation === BlueInThe.NORTH || orientation === BlueInThe.SOUTH
  }
}

export class CanvasOrient {
  height = 0
  width = 0
  orient = BlueInThe.EAST
}

export interface CanvasProps {
  orientation: CanvasOrient
  picType: string
  braaFirst: boolean
  dataStyle: SensorType
  showMeasurements: boolean
  isHardMode: boolean
  newPic: boolean
  animate: boolean
  resetCallback?: () => void
  animateCallback: () => void
}

export interface PictureCanvasState {
  bullseye: Point
  blueAir: AircraftGroup
  answer: PictureAnswer
  reDraw: PictureReDrawFunction
  animateCanvas?: ImageData
}

export interface CanvasDrawFunction {
  (context: CanvasRenderingContext2D): Promise<void>
}

export interface DrawCanvasProps extends CanvasProps {
  draw: CanvasDrawFunction
  bullseye: Point
  answer: PictureAnswer
}

export interface PictureCanvasProps extends CanvasProps {
  format: FORMAT
  setAnswer: { (answer: PictureAnswer): void }
  sliderSpeed: number
  desiredNumContacts: number
}

export type PictureAnswer = {
  pic: string
  groups: AircraftGroup[]
}

export interface PictureDrawFunction {
  (
    context: CanvasRenderingContext2D,
    props: PictureCanvasProps,
    state: PictureCanvasState,
    hasCaps: boolean,
    desiredNumContacts: number,
    start?: Point
  ): PictureAnswer
}

export interface PictureReDrawFunction {
  (forced?: boolean, start?: Point, hasCaps?: boolean): PictureAnswer
}
