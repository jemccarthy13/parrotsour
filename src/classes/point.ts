import { BlueInThe } from "../canvas/canvastypes"
import { BRAA } from "./braa"
import { PIXELS_TO_NM, toDegrees } from "../utils/psmath"

/**
 * An x,y Cartesian point that can calculate bearing and
 * range to another point.
 */
export class Point {
  x: number
  y: number

  static DEFAULT = new Point(0, 0)

  /**
   * Construct a new x,y point
   */
  constructor(x: number, y: number) {
    this.x = x
    this.y = y
  }

  /**
   * Get a bearing and range between this point and another point
   *
   * @param ptTo the "other" or "to" point
   * @returns the bearing and range between this (ptFrom) and ptTo
   */
  getBR(toPoint: Point): BRAA {
    const deltaX = this.x - toPoint.x
    const deltaY = this.y - toPoint.y

    // distance formula for range
    const rng = Math.floor(
      Math.sqrt(deltaX * deltaX + deltaY * deltaY) / PIXELS_TO_NM
    )

    // convert cartesian angle to heading
    let brg = Math.round(
      270 + toDegrees(Math.atan2(this.y - toPoint.y, this.x - toPoint.x))
    )
    if (brg > 360) {
      brg = brg - 360
    }

    return new BRAA(brg, rng)
  }

  /**
   * Get distance between points (straight along axis, i.e. not slant range)
   */
  public straightDistNM(toPoint: Point, orientation: BlueInThe): number {
    return Math.floor(
      Math.abs(
        orientation === BlueInThe.NORTH || orientation === BlueInThe.SOUTH
          ? this.y - toPoint.y
          : this.x - toPoint.x
      ) / PIXELS_TO_NM
    )
  }
}
