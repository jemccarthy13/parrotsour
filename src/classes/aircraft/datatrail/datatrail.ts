import { Point } from "../../point"
import { IDMatrix } from "../id"

// TODO -- DATATRAIL -- register? so that Aircraft can generate one of each type of datatrail?
// TODO -- DATATRAIL -- Support a toggle for the different Sensor Types.....
export abstract class DataTrail {
  private startPos: Point

  constructor(startPos: Point) {
    this.startPos = startPos
  }

  public getStartPos(): Point {
    return this.startPos
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  public move(_heading: number): void {
    //console.warn("Moving DataTrail in base class. Will have no effect.")
  }

  abstract getCenterOfMass(heading: number): Point

  abstract draw(heading: number, id: IDMatrix): void
}
