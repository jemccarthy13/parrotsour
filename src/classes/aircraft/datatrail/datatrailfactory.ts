import { Point } from "../../point"
import { ArrowDataTrail } from "./arrowdatatrail"
import { DataTrail } from "./datatrail"
import { RawDataTrail } from "./rawdatatrail"
import { SensorType } from "./sensortype"

/**
 * Abstracts DataTrails from Aircraft - if a new DataTrail is added, it'll
 * add a new SensorType, a new constructor here, and implement the
 * required methods. Aircraft will automatically know to create that kind of trail.
 */
export class DataTrailFactory {
  // TODO - DATATRAIL -- self-registering DataTrail child classes, if possible
  // static function register(type, createFunction)
  // each child has create(startPos) => return new [subtype]DataTrail
  static create(type: SensorType, startPos: Point, heading: number): DataTrail {
    switch (type) {
      case SensorType.ARROW:
        return new ArrowDataTrail(startPos)
      case SensorType.RAW:
        return new RawDataTrail(startPos, heading)
    }
  }
}
