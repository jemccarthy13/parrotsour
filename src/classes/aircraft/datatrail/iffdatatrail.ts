import { PaintBrush } from "../../../canvas/draw/paintbrush"
import { headingToRadians, PIXELS_TO_NM } from "../../../utils/psmath"
import { Point } from "../../point"
import { IDMatrix } from "../id"
import { DataTrail } from "./datatrail"

export class IFFDataTrail extends DataTrail {
  private iffPoints: Point[] = []
  private LEN_TRAIL: number = 7 * PIXELS_TO_NM

  constructor(startPos: Point, heading: number) {
    super(startPos)

    const starty = startPos.y
    const startx = startPos.x
    const vector = headingToRadians(heading)
    const endy = starty + this.LEN_TRAIL * -Math.sin(vector.radians)
    const endx = startx + this.LEN_TRAIL * Math.cos(vector.radians)

    const offsetX = (endx - startx) / PIXELS_TO_NM
    const offsetY = (endy - starty) / PIXELS_TO_NM

    let xPos = startx
    let yPos = starty
    const iPts: Point[] = []
    // draw the radar trail
    if (!this.iffPoints || this.iffPoints.length === 0) {
      for (let mult = 0; mult < 5; mult++) {
        xPos = startx + offsetX * mult + offsetX * 0.5
        yPos = starty + offsetY * mult + offsetY * 0.5
        iPts.push(new Point(xPos, yPos))
      }
      this.iffPoints = iPts
    }
  }

  getCenterOfMass(): Point {
    return this._getOnePlotAhead()
  }

  public move(heading: number): void {
    // this logic should be in move()
    // take the first point out of the radar trail
    const newIFFPts = this.iffPoints.slice(1)

    const startPos = this.getStartPos()
    const startx = startPos.x
    const starty = startPos.y
    const vector = headingToRadians(heading)

    const endy = starty + this.LEN_TRAIL * -Math.sin(vector.radians)
    const endx = startx + this.LEN_TRAIL * Math.cos(vector.radians)
    const offsetX = (endx - startx) / PIXELS_TO_NM
    const offsetY = (endy - starty) / PIXELS_TO_NM

    // add a bit of jitter with randomness
    const jit = 1

    const xPos = startx + offsetX * 6 + jit * Math.random() + Math.random()
    const yPos = starty + offsetY * 6 + jit * Math.random() + Math.random()
    newIFFPts.push(new Point(xPos, yPos))
    this.iffPoints = newIFFPts
  }

  draw(heading: number, id: IDMatrix): void {
    const ctx = PaintBrush.getContext()
    if (id !== IDMatrix.SUSPECT && id !== IDMatrix.HOSTILE) {
      ctx.strokeStyle = "blue"
      ctx.lineWidth = 1
      for (let l = 0; l < this.iffPoints.length; l++) {
        const xPos = this.iffPoints[l].x
        const yPos = this.iffPoints[l].y
        ctx.beginPath()
        ctx.moveTo(xPos, yPos)
        ctx.lineTo(xPos - 3, yPos)
        ctx.lineTo(xPos - 3, yPos - 3)
        ctx.lineTo(xPos, yPos - 3)
        ctx.lineTo(xPos, yPos)
        ctx.stroke()
      }
    }
  }

  _getOnePlotAhead(): Point {
    const cPt = this.iffPoints[this.iffPoints.length - 1]
    const pPt = this.iffPoints[this.iffPoints.length - 2]

    const deltX = cPt.x - pPt.x
    const deltY = cPt.y - pPt.y

    return new Point(cPt.x + deltX, cPt.y + deltY)
  }
}
