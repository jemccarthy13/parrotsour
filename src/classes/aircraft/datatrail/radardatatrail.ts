import { PaintBrush } from "../../../canvas/draw/paintbrush"
import {
  headingToRadians,
  PIXELS_TO_NM,
  toRadians,
} from "../../../utils/psmath"
import { Point } from "../../point"
import { IDMatrix } from "../id"
import { DataTrail } from "./datatrail"

export class RadarDataTrail extends DataTrail {
  private radarPoints: Point[] = []

  public static LEN_TRAIL: number = 7 * PIXELS_TO_NM

  constructor(startPos: Point, heading: number) {
    super(startPos)

    const starty = startPos.y
    const startx = startPos.x
    const vector = headingToRadians(heading)
    const endy = starty + RadarDataTrail.LEN_TRAIL * -Math.sin(vector.radians)
    const endx = startx + RadarDataTrail.LEN_TRAIL * Math.cos(vector.radians)

    const offsetX = (endx - startx) / PIXELS_TO_NM
    const offsetY = (endy - starty) / PIXELS_TO_NM

    let xPos = startx
    let yPos = starty
    const rdrPts: Point[] = []
    // draw the radar trail
    if (!this.radarPoints || this.radarPoints.length === 0) {
      for (let mult = 0; mult < 6; mult++) {
        // add a bit of jitter with randomness
        const jit = 3
        xPos = startx + offsetX * mult + jit * Math.random() + Math.random()
        yPos = starty + offsetY * mult + jit * Math.random() + Math.random()
        const rdrPt = new Point(xPos, yPos)
        rdrPts.push(rdrPt)
      }
      this.radarPoints = rdrPts
    }
  }

  getCenterOfMass(): Point {
    return this._getOnePlotAhead()
  }

  public move(heading: number): void {
    // this logic should be in move()
    // take the first point out of the radar trail
    const newRdrPts = this.radarPoints.slice(1)

    const startPos = this.getStartPos()
    const startx = startPos.x
    const starty = startPos.y
    const vector = headingToRadians(heading)

    const endy = starty + RadarDataTrail.LEN_TRAIL * -Math.sin(vector.radians)
    const endx = startx + RadarDataTrail.LEN_TRAIL * Math.cos(vector.radians)
    const offsetX = (endx - startx) / PIXELS_TO_NM
    const offsetY = (endy - starty) / PIXELS_TO_NM

    // add a bit of jitter with randomness
    const jit = 3

    const xPos = startx + offsetX * 6 + jit * Math.random() + Math.random()
    const yPos = starty + offsetY * 6 + jit * Math.random() + Math.random()
    newRdrPts.push(new Point(xPos, yPos))
    this.radarPoints = newRdrPts
  }

  draw(heading: number, id: IDMatrix): void {
    const ctx = PaintBrush.getContext()
    // Draw radar dots
    ctx.strokeStyle = "#FF8C00"
    ctx.lineWidth = 2
    ctx.beginPath()
    this.radarPoints.forEach((pt) => {
      ctx.beginPath()
      ctx.moveTo(pt.x, pt.y)
      ctx.lineTo(pt.x - 4, pt.y - 4)
      ctx.stroke()
      ctx.stroke()
    })

    this._drawSymbology(ctx, heading, id)
  }

  public getDataTrail(): Point[] {
    return this.radarPoints
  }

  /**
   * @returns One anticipated data plot in front of the track
   */
  _getOnePlotAhead(): Point {
    const cPt = this.radarPoints[this.radarPoints.length - 1]
    const pPt = this.radarPoints[this.radarPoints.length - 2]

    const deltX = cPt.x - pPt.x
    const deltY = cPt.y - pPt.y

    return new Point(cPt.x + deltX, cPt.y + deltY)
  }

  /**
   * Draw Symbology for the track at the 'head' of the data trail
   * @param ctx Current drawing context
   * @param id ID for symbology (HOS/NEU/SUS/etc)
   */
  _drawSymbology(
    ctx: CanvasRenderingContext2D,
    heading: number,
    id: IDMatrix
  ): void {
    const plotAhead = this._getOnePlotAhead()

    ctx.strokeStyle = id

    if (id === IDMatrix.FRIEND) {
      // draw friend symbology (blue arc/upside-down U)
      ctx.moveTo(plotAhead.x - 2.5, plotAhead.y - 2.5)
      ctx.beginPath()
      ctx.arc(
        plotAhead.x - 2.5,
        plotAhead.y - 0.5,
        4,
        toRadians(170),
        toRadians(10)
      )
      ctx.stroke()
    } else {
      ctx.beginPath()
      const headX = plotAhead.x - 3
      const headY = plotAhead.y - 3

      const leftX = headX + 2 * PIXELS_TO_NM * Math.cos(toRadians(240))
      const leftY = headY + 2 * PIXELS_TO_NM - Math.sin(toRadians(240))
      ctx.moveTo(headX, headY)
      ctx.lineTo(leftX, leftY)

      const rightX = headX + 2 * PIXELS_TO_NM * Math.cos(toRadians(300))
      const rightY = headY + 2 * PIXELS_TO_NM - Math.sin(toRadians(300))
      ctx.moveTo(headX, headY)
      ctx.lineTo(rightX, rightY)
      ctx.stroke()
      ctx.stroke()
    }

    // Draw vector stick
    ctx.strokeStyle = "black"
    ctx.lineWidth = 1.5
    ctx.beginPath()

    const vector = headingToRadians(heading)
    const deltX = 3 * PIXELS_TO_NM * Math.cos(vector.radians)
    const deltY = 3 * PIXELS_TO_NM * -Math.sin(vector.radians)
    ctx.moveTo(plotAhead.x - 2.5, plotAhead.y - 2.5)
    ctx.lineTo(plotAhead.x - 2.5 + 1.2 * deltX, plotAhead.y - 2.5 + 1.2 * deltY)
    ctx.stroke()
    ctx.stroke()
  }
}
