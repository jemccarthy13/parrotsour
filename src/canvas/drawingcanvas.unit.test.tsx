import React from "react"
import { mount } from "enzyme"
import DrawingCanvas, { CanvasMouseEvent } from "./drawingcanvas"

import { Point } from "../classes/point"
import { BlueInThe, DrawCanvasProps } from "./canvastypes"
import { SensorType } from "../classes/aircraft/datatrail/sensortype"

// eslint-disable-next-line @typescript-eslint/no-var-requires
const canvasSerializer = require("jest-canvas-snapshot-serializer")
expect.addSnapshotSerializer(canvasSerializer)

/**
 * Mock draw function for a drawing canvas
 * @param context the Context to draw in
 */
const drawMock = async (
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  ctx: CanvasRenderingContext2D | null | undefined
): Promise<void> => {
  return new Promise(jest.fn())
}

const testProps: DrawCanvasProps = {
  draw: drawMock,
  bullseye: new Point(0, 0),
  orientation: { height: 400, width: 400, orient: BlueInThe.NORTH },
  picType: "azimith",
  braaFirst: true,
  dataStyle: SensorType.ARROW,
  newPic: false,
  animate: false,
  resetCallback: jest.fn(),
  animateCallback: jest.fn(),
  answer: { pic: "2 GRPS AZ", groups: [] },
  showMeasurements: true,
  isHardMode: false,
}

describe("drawingCanvas", () => {
  let drawingCanvas = mount(<DrawingCanvas {...testProps} />)

  it("renders", () => {
    expect(drawingCanvas).not.toBe(null)
  })

  it("mousemove_singular", () => {
    drawingCanvas = mount(<DrawingCanvas {...testProps} />)
    expect(drawingCanvas).not.toBe(null)

    const moveEvent: CanvasMouseEvent = {
      clientX: 50,
      clientY: 50,
      getModifierState: () => false,
    }

    const mouseCanvas = drawingCanvas.find("#mousecanvas")
    mouseCanvas.simulate("mousemove", { ...moveEvent })

    const mCanvasInstance = mouseCanvas.instance()
    expect(mCanvasInstance).toMatchSnapshot()
  })

  it("mousemove_multiple", () => {
    drawingCanvas = mount(<DrawingCanvas {...testProps} />)
    expect(drawingCanvas).not.toBe(null)

    const moveEvent: CanvasMouseEvent = {
      clientX: 50,
      clientY: 50,
      getModifierState: () => false,
    }
    const moveEvent2: CanvasMouseEvent = {
      clientX: 75,
      clientY: 50,
      getModifierState: () => false,
    }

    const mouseCanvas = drawingCanvas.find("#mousecanvas")
    mouseCanvas.simulate("mousemove", { ...moveEvent })
    mouseCanvas.simulate("mousemove", { ...moveEvent2 })

    const mCanvasInstance = mouseCanvas.instance()
    expect(mCanvasInstance).toMatchSnapshot()
  })

  it("click_and_drag", () => {
    drawingCanvas = mount(<DrawingCanvas {...testProps} />)
    expect(drawingCanvas).not.toBe(null)

    const moveEvent: CanvasMouseEvent = {
      clientX: 50,
      clientY: 50,
      getModifierState: () => false,
    }
    const moveEvent2: CanvasMouseEvent = {
      clientX: 75,
      clientY: 50,
      getModifierState: () => false,
    }

    const mouseCanvas = drawingCanvas.find("#mousecanvas")
    mouseCanvas.simulate("mousedown", { ...moveEvent })
    mouseCanvas.simulate("mousemove", { ...moveEvent2 })
    mouseCanvas.simulate("mouseup", { ...moveEvent2 })

    const mCanvasInstance = mouseCanvas.instance()
    expect(mCanvasInstance).toMatchSnapshot()
  })
})
