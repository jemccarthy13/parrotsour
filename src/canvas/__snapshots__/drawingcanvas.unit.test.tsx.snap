// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawingCanvas click_and_drag 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAZAAAAGQCAYAAACAvzbMAAAABmJLR0QA/wD/AP+gvaeTAAAE6klEQVR4nO3csYscZRgH4J+QwkJCCCEWIpI+SMBODo2NjeQPiGCXtCKkECHBwiIW/glXKyjYiDYGzJG0toZwhYgQQtIE0cTsmnst5jtvmOzOHupN2ON5YNnZb973m5lmX+Z7ZzcBAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA4DOpiUo+TqqT+SOpSG38nqd/b+J2kXhmPT5LaSGqe1PGkbiW1k9Qsqfenvy4ADljdSeqrpF5M6lpSd9v4b0l9ltSxpK4k9dN4fJLUZlLbSd1Iaiupl9ocs+mvC4ADVn/17i5e2/uyr+qKRJLUyS5uLD5pxWMzqUdJvTvN+QPwjFQl9XzbPtotOyVteepyuwP5sosbi0/a8tVGW7q6kdSTVkzOT3c9AEykdrpCkLRisVtAzrcv/52kvu+NL4tv/Y+kFZmv2/4Pu14KAIdMzZI63bbPJPXngpi3u2IyFr/b//gn5uW2fWLv7gWAQ6R+Tuqb9uTUt12TPGlF4GpSp5L6LqkfVsS3/keS1I9Jfd4tddXHXUMegEOmPmiN8Ur3eO7F3visLVndTuqFFfGt/5F0Tfb6teU+TOrc9NcFAAAAAACQt5I8GYxdSPIwSbX39wb7P2n7+jaSzEdyV80JwBr5NMksTxeDx+mKxLEkXyR51Nt3vH0e5mwm2R7JHZsTgDVzL8m5PF0M+k4k6f3iPNeTXFuQs52uiIzlrhoHYM2MFZCPkrQ/U8yb6e4cji7ImadbxlqWu59xANbMsgLyapL7Sdovy3M3yZUFObv9j7HcVeMArKFFBeRMkl+SvD6IG76Svf7HWO7YOABralhALid5kOT0PnP6/Y9lufuZE4D/6LmJj1eDY86THBnEDM+pnzNP9zjwzSW5W0neWDDH1r8836Gz/9M8AGtv6gJykK5PcIyzExwDAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAOxt/fG1tcVqC13gAAAABJRU5ErkJggg=="
  height="400"
  id="mousecanvas"
  style="background-color: transparent; width: 400px; height: 400px; border: 1px solid #000000; grid-column: 2; grid-row: 1; position: absolute; left: 0px;"
  width="400"
/>
`;

exports[`drawingCanvas mousemove_multiple 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAZAAAAGQCAYAAACAvzbMAAAABmJLR0QA/wD/AP+gvaeTAAADaUlEQVR4nO3ZP4rWUBTG4d8KRMTWHYhMLYK4AJdgN5sQnMrGRdi7iSlmFTZT2Yk2U80wDv4pvsiETJIvbeB5ICT33Ly3PSSnAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAC496b6PamdVtfV3+H+brL/cdgbe1XdrWSPnQnAjnyqfvWwGdx2aBKPqy/VzWjvybCeZj5XlyvZtTMB2Jkf1dseNoOxp9Wf0fqiOp/JXHZoImvZY3UAdmatgbyvvg7Przt8OTyaydx1+I21lN1SB2BnlhrIi+pn9WxYf6/OZjL/5x9r2WN1AHZoroGcVN+ql5P3plfdzz/Wsmt1AHZq2kA+VFfV842Z8fxjKbvlTAB2Zm6eMfelsZQZzz+WslvOBAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGCTf0xjXiO47mRRAAAAAElFTkSuQmCC"
  height="400"
  id="mousecanvas"
  style="background-color: transparent; width: 400px; height: 400px; border: 1px solid #000000; grid-column: 2; grid-row: 1; position: absolute; left: 0px;"
  width="400"
/>
`;

exports[`drawingCanvas mousemove_singular 1`] = `
<canvas
  data-snapshot-image="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAZAAAAGQCAYAAACAvzbMAAAABmJLR0QA/wD/AP+gvaeTAAADuElEQVR4nO3Zv2oVQRTA4R8WEnwBEfEFLOzsBLETREgt2lnYCmptaSloGVsfQbS08R0kLyAq4j8UEzQWu5DrNXejV4Ub+T7YYs/Mnj3VHGamAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAOinPV17nYevWh2qk+VjfG+JkxNvs0M7Y9kXP+u50AOLDuVFv9vJh/qh5Ux6rb7TaGjer5glwb1eZEzlkPq0dLVQzASnhZXWzxYn+ioTG8G983q3sL5m6Oc/fLeX7Md3iJegFYMXst9serb+PY3TG2Xb0dY++rCzPztxuOsaZyVr2obv1JsQCsjqnjpvsNR1Kz1qpLDfck9eP9x1TO9YbjMQD+E1MN5OiC8bWGHUrt3n/sl/NZ9eS3qwNgpRyaGPtSXa+OVDerV2N8q7rW0DwuV2/G+Nnq6S/881T1eJliAVhN87uFKw1HTTsNzeP0GL9afW7Yebyeic/ff+yVs/G7k3+hXgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD+ue8yWlPF0x5KPgAAAABJRU5ErkJggg=="
  height="400"
  id="mousecanvas"
  style="background-color: transparent; width: 400px; height: 400px; border: 1px solid #000000; grid-column: 2; grid-row: 1; position: absolute; left: 0px;"
  width="400"
/>
`;
